# Pandas dataframes

Pandas dataframes are mutable two-dimensional structures of data with labeled axes where: 
* each row represents a different observation
* each column represents a different variable

In Python, to define a dataframe, we first need to import the pandas module.

In [4]:
import pandas as pd

Next, if we want a dataframe with 5 rows and 2 columns, we can do it from a [dictionary](https://www.w3schools.com/python/python_dictionaries.asp), a [list](https://www.w3schools.com/python/python_lists.asp) of lists, a list of dictionaries, etc.

We are going to create a 5 row, 2 column dataframe from a dictionary. To do this, we will follow these steps:

1. Create a dictionary where the keys will be the names of the columns and the values will be lists, with as many elements as the number of rows we want.
2. Convert that dictionary to dataframe with pandas' DataFrame() function

In [5]:
# 1. Create the dictionary
data = {
    "x":[1, 2, 3, 4, 5], 
    "y":[6, 7, 8, 9, 10],
    'Boolean data' : [True, False, True, True, False] 
}

In [9]:
data['x']
data['y']

[6, 7, 8, 9, 10]

In [6]:
# 2. Convert dictionary to dataframe
df = pd.DataFrame(data)
print(df)
# df

   x   y  Boolean data
0  1   6          True
1  2   7         False
2  3   8          True
3  4   9          True
4  5  10         False


As we said, we have created a dataframe with 5 rows and two columns, called x and y respectively.

**Observation**: As a result of `print()`, we have not only obtained the 5 rows and 2 columns, but there is an additional "column" of 5 numbers ordered vertically from 0 to 4. This column is called the **index** and it is simply the name of each row, which by default is their ordinal position: 0 indicates the first row; the 1, the second; and so on.

## Import a csv file to dataframe

Most of the time, you will not be creating dataframes yourself, but importing (or "reading") data from a csv file or a database into a pandas dataframe. It's easy to do with pandas' read functions:

In [7]:
# use contextual help to show all the parameters inside read_csv
df = pd.read_csv('orderlines.csv', sep=',')
df

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

## Dataframe dimensions

With the `.shape` [method](https://www.w3schools.com/python/gloss_python_object_methods.asp) we can calculate the dimensions (number of rows and columns) of the dataframe.

In [14]:
df.shape

(293983, 7)

As a result we obtain a [tuple](https://www.w3schools.com/python/python_tuples.asp) where the first element is the number of rows, which in our case is 293983, while the second element is the number of columns, which in our example was 7.

In [7]:
nrows = df.shape[0]
ncols = df.shape[1]
print("The number of rows is", nrows)
print("The number of columns is", ncols)

The number of rows is 293983
The number of columns is 7


`DataFrame.size` returns the total number of values that the dataframe has (number of rows per number of columns):

In [15]:
df.size

2057881

In [16]:
# check if that's true
df.shape[0] * df.shape[1] == df.size

True

With the `.ndim` method we calculate the number of dimensions that the dataframe has. This will always be 2, as it consists of rows and columns.

In [17]:
df.ndim

2

## Dataframes exploration

The `DataFrame.head()` and `DataFrame.tail()` methods are used to display the first or last rows of the dataframe. Looking at the raw data is a great way to get a grasp of what's in there. By default, 5 rows will be shown, but you can change that:

In [14]:
df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [15]:
df.head(9)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38  
5  2017-01-01 01:14:27  
6  2017-01-01 01:17:21  
7  2017-01-01 01:46:16  
8  2017-01-01 01:50:34

In [16]:
df.tail()

id  id_order  product_id  product_quantity      sku unit_price  \
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36

The methods `DataFrame.info()`, `DataFrame.describe()` and `DataFrame.nunique()` give a general overview of what's in the dataframe:

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [51]:
df.describe()

id       id_order  product_id  product_quantity
count  2.939830e+05  293983.000000    293983.0     293983.000000
mean   1.397918e+06  419999.116544         0.0          1.121126
std    1.530096e+05   66344.486479         0.0          3.396569
min    1.119109e+06  241319.000000         0.0          1.000000
25%    1.262542e+06  362258.500000         0.0          1.000000
50%    1.406940e+06  425956.000000         0.0          1.000000
75%    1.531322e+06  478657.000000         0.0          1.000000
max    1.650203e+06  527401.000000         0.0        999.000000

In [57]:
df['num_unit_price'] = pd.to_numeric(df['unit_price'], errors = 'coerce' , downcast = None)
df.head(10)
#df.unit_price.min()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   
9  1119120    295347           0                 1  APP0700      72.19   

                  date  num_unit_price  
0  2017-01-01 00:07:19           18.99  
1  2017-01-01 00:19:45          399.00  
2  2017-01-01 00:20:57          474.05  
3  2017-01-01 00:51:40           68.39  
4  2017-01-01 01:06:38           23.74  
5  2017-01-01 01:14:27          231.79  
6  2017-01-01 01:17:21             NaN  
7  2017-01-01 01:46:16           47.49  
8  2017-01-01 01:50:34           18.99  
9  2017-01-01 01:54:11           72.19

In [54]:
df.nunique()

id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
num_unit_price        7838
dtype: int64

The `.unique()` method returns the unique values from a column as a numpy array, which can be indexed with `[]`:

In [22]:
df['sku'].unique()[:10]

array(['OTT0133', 'LGE0043', 'PAR0071', 'WDT0315', 'JBL0104', 'WDT0249',
       'APP1582', 'OWC0100', 'IOT0014', 'APP0700'], dtype=object)

The `isna()` method returns a boolean for each value: `True` if that value is "missing" (which is represented as `NaN` in numpy and pandas) and `False` if the value is not missing:

In [23]:
df.isna()

id  id_order  product_id  product_quantity    sku  unit_price  \
0       False     False       False             False  False       False   
1       False     False       False             False  False       False   
2       False     False       False             False  False       False   
3       False     False       False             False  False       False   
4       False     False       False             False  False       False   
...       ...       ...         ...               ...    ...         ...   
293978  False     False       False             False  False       False   
293979  False     False       False             False  False       False   
293980  False     False       False             False  False       False   
293981  False     False       False             False  False       False   
293982  False     False       False             False  False       False   

         date  
0       False  
1       False  
2       False  
3       False  
4       False  
...       ...  
293978  False  
293979  False  
293980  False  
293981  False  
293982  False  

[293983 rows x 7 columns]

We can then use `DataFrame.sum()` to add up all these booleans for each column, and count how many missing values are there in the dataframe, since `True` is interpreted as `1` and `False` as `0`:

In [33]:
df.isna().sum()

id                      0
id_order                0
product_id              0
product_quantity        0
sku                     0
unit_price              0
date                    0
num_unit_price      36169
dtype: int64

`DataFrame.duplicated()` also returns a boolean output, but in this case just one value per row: `Ture` if that row is duplicated and `False` if it's not. Again, using `sum()` allows us to count how many `True` values (i.e. duplicated rows) are there in total:

In [34]:
df.duplicated().sum() # parameters keep=False
# df.drop_duplicates()

0

In [36]:
df.drop_duplicates().shape

(293983, 8)

`DataFrame.nlargest(n, columns)` will return the top `n` rows with the largest value for whatever column we specify in `columns`. Below, we see the rows with the largest product quantity values:

In [50]:
df.nlargest(5, 'product_quantity')

id  id_order  product_id  product_quantity      sku unit_price  \
53860  1228150    346221           0               999  APP1190      55.99   
68712  1254032    358747           0               999  SEV0028      19.99   
57796  1234924    349475           0               800  KIN0137       7.49   
57306  1234111    349133           0               555  APP0665      70.99   
40813  1204788    335057           0               201  THU0029      80.99   

                      date  
53860  2017-04-14 21:50:52  
68712  2017-05-24 14:51:58  
57796  2017-04-25 09:59:00  
57306  2017-04-24 10:20:13  
40813  2017-03-14 15:25:53

 `DataFrame.nsmallest()` does the same, for the smallest values:

In [68]:
df.nsmallest(5, 'product_quantity')
df.head(10)
df_wobooks = df.loc[df.sku != 'LIBRO']
#df_wobooks
df_wobooks.nsmallest(55, 'num_unit_price')
#df_wobooks.where(df_wobooks.num_unit_price>2).nsmallest(5,'num_unit_price')


id  id_order  product_id  product_quantity        sku unit_price  \
77008   1268645    365886           0                 1    APP1465    -119.00   
53515   1227566    345934           0                 1  KIN0153-2       0.00   
53530   1227590    345957           0                 1    WDT0347       0.00   
70545   1257388    360246           0                 1    ENV0496       0.00   
71694   1259481    361198           0                 1    ENV0505       0.00   
94064   1299125    380508           0                 1    APP2264       0.00   
96747   1304078    382803           0                 1    APP1632       0.00   
101831  1313304    387198           0                 1    APP1020       0.00   
151297  1419235    430636           0                 1    LGE0044       0.00   
151410  1419460    430738           0                 1    APP1215       0.00   
151411  1419461    430738           0                 1    APP1214       0.00   
151412  1419462    430738           0                 1    LGE0062       0.00   
151413  1419463    430739           0                 1    APP1215       0.00   
151414  1419464    430739           0                 1    APP1214       0.00   
151415  1419465    430739           0                 1    LGE0062       0.00   
152311  1420637    431297           0                 1  APP0430-A       0.00   
152312  1420638    431297           0                 1    APP0697       0.00   
152314  1420640    431297           0                 1    SAT0047       0.00   
152315  1420641    431297           0                 1    LGE0063       0.00   
152929  1421543    431730           0                 1    SEV0012       0.00   
153213  1422043    431992           0                 1    CRU0031       0.00   
153215  1422045    431994           0                 1    CRU0031       0.00   
153217  1422051    431998           0                 1    SYN0176       0.00   
153477  1422479    432195           0                 1    APP1216       0.00   
153478  1422480    432195           0                 1    APP2092       0.00   
153479  1422481    432195           0                 1    KAN0013       0.00   
153807  1423001    432467           0                 1    APP1979       0.00   
153808  1423002    432467           0                 1    APP1020       0.00   
153809  1423003    432467           0                 1    BOS0058       0.00   
153860  1423115    432509           0                 1    SHE0044       0.00   
153861  1423116    432509           0                 1    APP2558       0.00   
153862  1423117    432509           0                 1    APP2557       0.00   
153864  1423122    432510           0                 1    SHE0044       0.00   
153865  1423123    432510           0                 1    APP2558       0.00   
153866  1423124    432510           0                 1    APP2557       0.00   
153939  1423260    432575           0                 1    PAC2121       0.00   
154349  1424054    432940           0                 1    PAC1557       0.00   
154350  1424055    432940           0                 1    SEV0026       0.00   
154351  1424056    432940           0                 1    DLL0015       0.00   
154352  1424057    432940           0                 1    NTE0098       0.00   
154566  1424456    433109           0                 1    PAC2122       0.00   
154567  1424457    433109           0                 1    LAC0181       0.00   
154568  1424458    433109           0                 1    WAC0101       0.00   
154569  1424459    433110           0                 1    PAC2122       0.00   
154570  1424460    433110           0                 1    LAC0181       0.00   
154571  1424461    433110           0                 1    WAC0101       0.00   
154572  1424462    433111           0                 1    PAC2122       0.00   
154573  1424463    433111           0                 1    LAC0181       0.00   
154574  1424464    433111           0       

## Columns

Given a dataframe, we can select a particular column in several ways:

* Indicating the name of the column between square brackets, `[]`
* With the `.loc[]` method (by name or tag)
* With the `.iloc[]` method (by position)

The simple `[]` method is used to just view a column if you now its name and **don't want to modify it**:

In [70]:
# select the column by name
df['id_order']

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

The `.loc[]` takes two arguments: `[rows, columns]`. Passing `:` in the rows argument means "grabbing all the rows", which allows you to select a whole column if you know its name. This method is more flexible as you will see in the future, and would allow you to modify the data if needed.

In [71]:
# method .loc[]
df.loc[:, ['id_order', 'sku']]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

The `iloc[]` method works similarly, but only accept integers (which represent the positions of the rows and columns):

In [77]:
# method .iloc[]
print(df.iloc[:, 0])

0         1119109
1         1119110
2         1119111
3         1119112
4         1119113
           ...   
293978    1650199
293979    1650200
293980    1650201
293981    1650202
293982    1650203
Name: id, Length: 293983, dtype: int64


### Select multiple columns

If we wanted to select more than one column, we could do it with all the options listed above, with slight modifications in some cases:

In [27]:
# note that we pass a list inside of the []
df[['id_order','sku']]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

In [30]:
# .loc()
df.loc[:, ["id_order", "sku"]]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

With `.loc[]` and `:` you can select all columns between two columns you specify.

In [31]:
# .loc()
df.loc[:, "id_order":"sku"]

id_order  product_id  product_quantity      sku
0         299539           0                 1  OTT0133
1         299540           0                 1  LGE0043
2         299541           0                 1  PAR0071
3         299542           0                 1  WDT0315
4         299543           0                 1  JBL0104
...          ...         ...               ...      ...
293978    527398           0                 1  JBL0122
293979    527399           0                 1  PAC0653
293980    527400           0                 2  APP0698
293981    527388           0                 1  BEZ0204
293982    527401           0                 1  APP0927

[293983 rows x 4 columns]

In [32]:
# .iloc
df.iloc[:, [0, 1]]

id  id_order
0       1119109    299539
1       1119110    299540
2       1119111    299541
3       1119112    299542
4       1119113    299543
...         ...       ...
293978  1650199    527398
293979  1650200    527399
293980  1650201    527400
293981  1650202    527388
293982  1650203    527401

[293983 rows x 2 columns]

In [96]:
df.iloc[:, 0:2]

id  id_order
0       1119109    299539
1       1119110    299540
2       1119111    299541
3       1119112    299542
4       1119113    299543
...         ...       ...
293978  1650199    527398
293979  1650200    527399
293980  1650201    527400
293981  1650202    527388
293982  1650203    527401

[293983 rows x 2 columns]

## Rows

Given a dataframe, we can select a particular row in several ways:

* With the `.loc[]` method (by name or tag)
* With the `.iloc[]` method (by position)

Selecting a single row returns a Pandas Series (the 1-dimensional object that pandas has):

In [97]:
df.loc[0]

id                              1119109
id_order                         299539
product_id                            0
product_quantity                      1
sku                             OTT0133
unit_price                        18.99
date                2017-01-01 00:07:19
Name: 0, dtype: object

With `.loc[]`, rows are selected by its index name:

In [98]:
df.loc[0:3]

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40

If we change the index and set it to the `id` column, now the first rows can not be selected the same way:

In [99]:
df.set_index('id')

# inplace = True is the same as doing df = df.set_index('id'), i.e. it modifies the dataframe

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   
1119113    299543           0                 1  JBL0104      23.74   
...           ...         ...               ...      ...        ...   
1650199    527398           0                 1  JBL0122      42.99   
1650200    527399           0                 1  PAC0653     141.58   
1650201    527400           0                 2  APP0698       9.99   
1650202    527388           0                 1  BEZ0204      19.99   
1650203    527401           0                 1  APP0927      13.99   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45  
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40  
1119113  2017-01-01 01:06:38  
...                      ...  
1650199  2018-03-14 13:57:25  
1650200  2018-03-14 13:57:34  
1650201  2018-03-14 13:57:41  
1650202  2018-03-14 13:58:01  
1650203  2018-03-14 13:58:36  

[293983 rows x 6 columns]

In [100]:
df.iloc[:]

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

In [ ]:
df.iloc[0:3]

In [51]:
df.head(2)

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45

In [52]:
# select the first observation with the .loc() method
df.loc[1119110]

id_order                         299540
product_id                            0
product_quantity                      1
sku                             LGE0043
unit_price                       399.00
date                2017-01-01 00:19:45
Name: 1119110, dtype: object

With the `iloc[]` method we don't need to know the row names to select rows at a certain position:

In [40]:
df.iloc[0:3]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   
1119111    299541           0                 1  PAR0071     474.05   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45  
1119111  2017-01-01 00:20:57

In [101]:
# select the last observation with the method .iloc[]
df.iloc[-1]
# df.tail(1)

id                              1650203
id_order                         527401
product_id                            0
product_quantity                      1
sku                             APP0927
unit_price                        13.99
date                2018-03-14 13:58:36
Name: 293982, dtype: object

In [40]:
df.loc[[1119111,1119112,1119113]]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   
1119113    299543           0                 1  JBL0104      23.74   

                        date  
id                            
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40  
1119113  2017-01-01 01:06:38

Again, using `:` you can select all elements between the two that you specified:

In [41]:
df.loc[1119111:1119113]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   
1119113    299543           0                 1  JBL0104      23.74   

                        date  
id                            
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40  
1119113  2017-01-01 01:06:38

In [43]:
df.reset_index(inplace=True)

Indexing can get tricky sometimes, it's ok to take some time to get used to the methods we presented, and it's ok to have some trouble selecting the rows and columns you need. For an exhaustive guide on Pandas indexing, check out this link: https://pandas.pydata.org/docs/user_guide/indexing.html#indexing 

## Drop and Filter data

The `.drop()` method allows us to delete the rows or columns that we indicate.

**Note:** Again, if we want to directly apply the changes to the original dataframe, we need to indicate `inplace = True`. Otherwise, we are getting as an output just a "view" of how the dataframe looks like after the drop, but the original dataframe remains untouched.

In [202]:
# droping the column "unit_price"
# axis=1 means we want to drop a column, not a row (for rows, axis=0)
df.drop(['unit_price'], axis=1)
#print(df.drop(['unit_price'], axis=1).shape)
#print(df.drop(['unit_price'], axis='columns').shape)
#print(df.drop(columns=['unit_price']).shape)

id  id_order  product_id  product_quantity      sku  \
0       1119109    299539           0                 1  OTT0133   
1       1119110    299540           0                 1  LGE0043   
2       1119111    299541           0                 1  PAR0071   
3       1119112    299542           0                 1  WDT0315   
4       1119113    299543           0                 1  JBL0104   
...         ...       ...         ...               ...      ...   
293978  1650199    527398           0                 1  JBL0122   
293979  1650200    527399           0                 1  PAC0653   
293980  1650201    527400           0                 2  APP0698   
293981  1650202    527388           0                 1  BEZ0204   
293982  1650203    527401           0                 1  APP0927   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 6 columns]

How to filter information on a dataframe.

In [103]:
# rows with product quantity larger than 100
df[df['product_quantity'] > 100]

id  id_order  product_id  product_quantity      sku unit_price  \
27779   1180010    323959           0               126  ADN0039      34.99   
40813   1204788    335057           0               201  THU0029      80.99   
53860   1228150    346221           0               999  APP1190      55.99   
57306   1234111    349133           0               555  APP0665      70.99   
57796   1234924    349475           0               800  KIN0137       7.49   
68712   1254032    358747           0               999  SEV0028      19.99   
136675  1388261    417536           0               200  TRK0009      29.99   
204637  1500715    464858           0               192  APP1662     519.00   
246048  1574262    496172           0               164  EVU0013      19.99   
285492  1637611    522075           0               125  XDO0047      25.99   

                       date  
27779   2017-02-14 10:21:12  
40813   2017-03-14 15:25:53  
53860   2017-04-14 21:50:52  
57306   2017-04-24 10:20:13  
57796   2017-04-25 09:59:00  
68712   2017-05-24 14:51:58  
136675  2017-10-25 15:02:39  
204637  2017-12-17 15:53:04  
246048  2018-01-22 16:14:42  
285492  2018-03-06 10:07:54

The `.query()` method can be useful for this purpose, as it resembles SQL syntax. Note that it works only when the column values do not contain blank spaces. You can use any **Python Comparison Operators** you want inside the query method (find more information on this [link](https://www.w3schools.com/python/python_operators.asp)).

In [104]:
df.query('product_quantity > 100')

id  id_order  product_id  product_quantity      sku unit_price  \
27779   1180010    323959           0               126  ADN0039      34.99   
40813   1204788    335057           0               201  THU0029      80.99   
53860   1228150    346221           0               999  APP1190      55.99   
57306   1234111    349133           0               555  APP0665      70.99   
57796   1234924    349475           0               800  KIN0137       7.49   
68712   1254032    358747           0               999  SEV0028      19.99   
136675  1388261    417536           0               200  TRK0009      29.99   
204637  1500715    464858           0               192  APP1662     519.00   
246048  1574262    496172           0               164  EVU0013      19.99   
285492  1637611    522075           0               125  XDO0047      25.99   

                       date  
27779   2017-02-14 10:21:12  
40813   2017-03-14 15:25:53  
53860   2017-04-14 21:50:52  
57306   2017-04-24 10:20:13  
57796   2017-04-25 09:59:00  
68712   2017-05-24 14:51:58  
136675  2017-10-25 15:02:39  
204637  2017-12-17 15:53:04  
246048  2018-01-22 16:14:42  
285492  2018-03-06 10:07:54

The `isin()` method is very useful to find rows that match any of the values you have in a list. For example, here we are searching for rows where its `sku` matches any of the 2 sku's we listed:

In [105]:
# find out a column that contains a list
df['sku'].isin(['JBL0104', 'ADN0039'])

0         False
1         False
2         False
3         False
4          True
          ...  
293978    False
293979    False
293980    False
293981    False
293982    False
Name: sku, Length: 293983, dtype: bool

This expression can be used inside of `[]` or `.loc[]` to filter the rows that have a `True` value. This is called "boolean indexing" and it is really useful:

In [112]:
only_valid_skus = df['sku'].isin(['JBL0104', 'ADN0039', 'LIBRO'])
df.loc[only_valid_skus, :]



id  id_order  product_id  product_quantity      sku unit_price  \
4       1119113    299543           0                 1  JBL0104      23.74   
716     1120463    300212           0                 2  JBL0104      26.99   
774     1120607    300272           0                 1  JBL0104      24.99   
866     1120758    300349           0                 1  JBL0104      24.99   
1411    1122014    300851           0                 1  JBL0104      24.99   
...         ...       ...         ...               ...      ...        ...   
258180  1593360    503247           0                 1  JBL0104      23.99   
263790  1601885    506553           0                 1  JBL0104      24.99   
266635  1606326    508345           0                 2  JBL0104      24.99   
268446  1608949    509377           0                 1  JBL0104      24.99   
279801  1628218    517916           0                 1  JBL0104      27.99   

                       date  
4       2017-01-01 01:06:38  
716     2017-01-02 01:44:05  
774     2017-01-02 08:57:07  
866     2017-01-02 10:41:29  
1411    2017-01-02 18:50:13  
...                     ...  
258180  2018-01-31 20:55:50  
263790  2018-02-05 22:41:40  
266635  2018-02-08 19:20:21  
268446  2018-02-10 18:02:37  
279801  2018-02-26 20:51:30  

[360 rows x 7 columns]

Pandas compresses large outputs in Jupyter Notebooks. If you want to see more rows, you can change the options:

In [113]:
pd.options.display.max_rows = 100
# after running this cell, run the code above again and you will see all the rows.

For a complete list of all settings and options that can be tweaked, check out this: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

#### Modifying a dataframe & the `.copy()` method

Let's take a small chunk of data from our dataframe:

In [114]:
sample = df.iloc[:3,:]
sample

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

Now we pick a single cell from the sample we took and we assign a new value to it. A warning already tells us that this is a risky thing to do:

In [116]:
# we change the "unit_price" of the first row:
sample.iloc[0,4] = 'NEW VALUE HERE'

We can see the new value on the `sample` we took:

In [117]:
sample

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

..and, maybe to your surprise, we can see that the new value is also present on the original `df`!

In [118]:
df.head()

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   
3  1119112    299542           0                 1         WDT0315      68.39   
4  1119113    299543           0                 1         JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [183]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


When you take a chunk of data using `.loc[]` or `iloc[]` and assign it to a new object, the new object is just a "tag" pointing to the very same data as the original dataframe points to. We can avoid this using the method `.copy()`

In [120]:
df = pd.read_csv('orderlines.csv')
sample = df.iloc[:3,:].copy()
sample.iloc[0,4] = 'NEW VALUE HERE'
sample

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

In [121]:
df.head(3)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

As you can see, now it has not been modified.

# CHALLENGES

1. How many different unit prices does the product JBL0104 have? Combine a pandas filter method with the method `.nunique()`.

In [253]:
# your code here
only_valid_skus = df['sku'].isin(['JBL0104'])
df.loc[only_valid_skus, :].unit_price.nunique()

7

In [254]:
df.loc[df['sku'].isin(['JBL0104']), :]['unit_price'].value_counts()

24.99    20
23.74    15
26.99     8
23.99     8
22.99     5
22.31     1
27.99     1
Name: unit_price, dtype: int64

2. List the (unique) items that were sold in the order with the id_order 385921.

In [234]:
valid_order = df['id_order'].isin([385921])
df.loc[valid_order, :].sku.unique()

#df.loc[df['id_order'].isin([385921]), :]



array(['APP2431', 'APP2348', 'APP2131', 'APP1630', 'APP1735', 'APP1216',
       'APP2092', 'APP1215', 'ELA0017', 'MIN0010', 'ELA0039', 'BEA0046',
       'BOS0034', 'BEA0071', 'ELA0029', 'APP2161', 'HOC0008', 'NOM0026',
       'NOM0014'], dtype=object)

3. Consider the products with the sku's APP2431 and APP2348. Find out in how many orders were they present.

In [10]:
valid_sku = df['sku'].isin(['APP2431', 'APP2348'])
df.loc[valid_sku, :].id_order.nunique()

179

In [9]:
valid_sku = df['sku'].isin(['APP2431', 'APP2348'])
df.loc[valid_sku, :].id_order.value_counts()

379659    2
401302    2
382460    2
388422    2
402265    2
         ..
430641    1
430899    1
431344    1
431577    1
524949    1
Name: id_order, Length: 179, dtype: int64

In [13]:
df.loc[df['sku'].isin(['APP2431', 'APP2348']), :].sku.value_counts()

APP2348    119
APP2431     70
Name: sku, dtype: int64

In [23]:
# your code here
df.where
df.where((df.sku == 'APP2431') | (df.sku == 'APP2348')).id_order.nunique()

179

4. Create a new dataframe, `df_50`, with all the rows that have a product quantity higher than 500 and only the columns `id`, id `order`, `product_quantity` and `sku`. Be sure to use the method `.copy()`. Once the new dataframe is created, modify the column 'product_quantity' to 'quantity', and 'sku' to 'product_code'. To do so, you can use the methods `.rename()` or `.columns`.


In [14]:
# your code here
df_50 = df.query('product_quantity > 500')
df_50 = df_50.drop(['unit_price', 'date', 'product_id'], axis=1)
df_50 = df_50.rename(columns={"product_quantity": "quantity" , 'sku' : 'product_code'})
df_50

id  id_order  quantity product_code
53860  1228150    346221       999      APP1190
57306  1234111    349133       555      APP0665
57796  1234924    349475       800      KIN0137
68712  1254032    358747       999      SEV0028

5. Filter all the order lines where the product `XDO0047` has appeared. Sort the rows by their product quantity using the [`.sort_values()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html) from pandas, in a DESCENDING order. Then look at the main descriptive information of this results with the method `.describe()`.

In [251]:
# your code her
special_order = df['sku'].isin(['XDO0047'])
df.loc[special_order, :].sort_values('product_quantity',ascending=False).describe()


id       id_order  product_id  product_quantity
count  3.600000e+01      36.000000        36.0         36.000000
mean   1.485767e+06  458640.277778         0.0          4.472222
std    8.770532e+04   36891.598431         0.0         20.662576
min    1.365515e+06  406387.000000         0.0          1.000000
25%    1.405830e+06  425462.750000         0.0          1.000000
50%    1.471428e+06  453247.000000         0.0          1.000000
75%    1.564540e+06  492406.000000         0.0          1.000000
max    1.640821e+06  523533.000000         0.0        125.000000

In [1]:
pwd

'C:\\Users\\Ajith\\DataScience_python\\Eniac Project II'

OBSERVATION: as you may have noticed, the `unit_price` column is not being detected as a float number on pandas. Some prices have been corrupted. During this week we will discover how to fix them, but for the moment it will be enough to filter the product we would like to analyise, bring the data into a new dataframe and change the data type of the price column to float.